File: openrag/core/models/evaluation.py
"""Domain models for the evaluation feature.

An *evaluation dataset* pairs a corpus (the files to index) with a test set
(the questions to ask). An *evaluation run* indexes that corpus into a
throwaway partition, replays the test set against the live API through
promptfoo, and records three families of metrics: indexing speed, retrieval
quality, and answer quality.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum

#: Runs index their corpus into a throwaway partition named from the run id.
#: These are an implementation detail of an eval and are filtered out of the
#: partition listings so they never show up as user-facing collections.
EVAL_PARTITION_PREFIX = "__eval_"


def is_eval_partition(partition: str) -> bool:
    """True for the throwaway partition a run creates."""
    return partition.startswith(EVAL_PARTITION_PREFIX)


class EvalRunStatus(str, Enum):
    """Lifecycle of a single evaluation run.

    Mirrors the indexing task vocabulary (``services.workers.task_state``) so
    the admin UI can reuse the same status styling.
    """

    QUEUED = "QUEUED"
    INDEXING = "INDEXING"
    EVALUATING = "EVALUATING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"

    @property
    def is_terminal(self) -> bool:
        return self in (EvalRunStatus.COMPLETED, EvalRunStatus.FAILED, EvalRunStatus.CANCELLED)


@dataclass(frozen=True)
class EvalTestCase:
    """One row of the uploaded test set CSV.

    ``expected_file_ids`` is optional: rows without it still contribute to the
    answer-quality metrics, but are excluded from hit rate / MRR / recall
    rather than being counted as misses.
    """

    query: str
    expected_answer: str
    expected_file_ids: tuple[str, ...] = ()

    @property
    def has_ground_truth_sources(self) -> bool:
        return bool(self.expected_file_ids)


@dataclass
class EvalDataset:
    """A stored corpus + test set pair."""

    id: str
    name: str
    corpus_file_count: int
    testset_row_count: int
    created_at: datetime | None = None
    created_by: int | None = None


@dataclass
class FileIndexingSample:
    """Wall-clock cost of indexing one corpus file."""

    filename: str
    size_bytes: int
    duration_seconds: float
    failed: bool = False


@dataclass
class IndexingMetrics:
    """Aggregate indexing speed over a run's corpus."""

    files_total: int = 0
    files_failed: int = 0
    bytes_total: int = 0
    wall_seconds: float = 0.0
    files_per_minute: float = 0.0
    megabytes_per_second: float = 0.0
    p50_seconds: float = 0.0
    p95_seconds: float = 0.0
    by_extension: dict[str, dict[str, float]] = field(default_factory=dict)
    samples: list[FileIndexingSample] = field(default_factory=list)


@dataclass
class RetrievalMetrics:
    """Ranking quality of the retrieved chunks.

    ``scored_cases`` counts the test rows that carried ``expected_file_ids``;
    ``skipped_cases`` counts those that did not. Both are reported so a
    near-empty ground truth can never masquerade as a perfect score.
    """

    scored_cases: int = 0
    skipped_cases: int = 0
    hit_rate: float = 0.0
    mrr: float = 0.0
    recall: float = 0.0
    context_relevance: float | None = None


@dataclass
class AnswerMetrics:
    """LLM-graded quality of the generated answers."""

    scored_cases: int = 0
    pass_rate: float = 0.0
    factuality: float | None = None
    rubric_score: float | None = None


@dataclass
class EvalCaseResult:
    """Per-question detail surfaced in the run detail table."""

    query: str
    retrieved_file_ids: list[str] = field(default_factory=list)
    expected_file_ids: list[str] = field(default_factory=list)
    hit: bool | None = None
    reciprocal_rank: float | None = None
    answer: str | None = None
    answer_passed: bool | None = None
    grader_reason: str | None = None


@dataclass
class EvalRun:
    """One evaluation execution against a dataset."""

    id: str
    dataset_id: str
    status: EvalRunStatus = EvalRunStatus.QUEUED
    started_at: datetime | None = None
    finished_at: datetime | None = None
    indexing: IndexingMetrics | None = None
    retrieval: RetrievalMetrics | None = None
    answer: AnswerMetrics | None = None
    cases: list[EvalCaseResult] = field(default_factory=list)
    error: str | None = None
    created_by: int | None = None


__all__ = [
    "EVAL_PARTITION_PREFIX",
    "AnswerMetrics",
    "EvalCaseResult",
    "EvalDataset",
    "EvalRun",
    "EvalRunStatus",
    "EvalTestCase",
    "FileIndexingSample",
    "IndexingMetrics",
    "RetrievalMetrics",
    "is_eval_partition",
]

File: ui/src/lib/api/evaluation.ts
import { request } from "./client";

// Admin evaluation endpoints:
//   GET    /evaluation/datasets          → EvalDataset[]
//   POST   /evaluation/datasets          → EvalDataset (multipart)
//   DELETE /evaluation/datasets/{id}     → 204
//   GET    /evaluation/runs              → EvalRunSummary[]
//   POST   /evaluation/runs              → EvalRun (202, 409 when one is in flight)
//   GET    /evaluation/runs/{id}         → EvalRun
//   POST   /evaluation/runs/{id}/cancel  → EvalRun

export interface EvalDataset {
  id: string;
  name: string;
  corpus_file_count: number;
  testset_row_count: number;
  created_at: string | null;
  created_by: number | null;
}

export type EvalRunStatus =
  | "QUEUED"
  | "INDEXING"
  | "EVALUATING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

export const ACTIVE_RUN_STATUSES: EvalRunStatus[] = ["QUEUED", "INDEXING", "EVALUATING"];

export function isActiveStatus(status: EvalRunStatus): boolean {
  return ACTIVE_RUN_STATUSES.includes(status);
}

/** Refetch cadence while a run can still change, shared by both eval views. */
export const EVAL_POLL_MS = 3000;

export interface FileIndexingSample {
  filename: string;
  size_bytes: number;
  duration_seconds: number;
  failed: boolean;
}

export interface IndexingMetrics {
  files_total: number;
  files_failed: number;
  bytes_total: number;
  wall_seconds: number;
  files_per_minute: number;
  megabytes_per_second: number;
  p50_seconds: number;
  p95_seconds: number;
  by_extension: Record<string, Record<string, number>>;
  samples: FileIndexingSample[];
}

export interface RetrievalMetrics {
  scored_cases: number;
  skipped_cases: number;
  hit_rate: number;
  mrr: number;
  recall: number;
  context_relevance: number | null;
}

export interface AnswerMetrics {
  scored_cases: number;
  pass_rate: number;
  factuality: number | null;
  rubric_score: number | null;
}

export interface EvalCase {
  query: string;
  retrieved_file_ids: string[];
  expected_file_ids: string[];
  hit: boolean | null;
  reciprocal_rank: number | null;
  answer: string | null;
  answer_passed: boolean | null;
  grader_reason: string | null;
}

export interface EvalRun {
  id: string;
  dataset_id: string;
  status: EvalRunStatus;
  started_at: string | null;
  finished_at: string | null;
  indexing: IndexingMetrics | null;
  retrieval: RetrievalMetrics | null;
  answer: AnswerMetrics | null;
  cases: EvalCase[];
  error: string | null;
  created_by: number | null;
}

export interface EvalRunSummary {
  id: string;
  dataset_id: string;
  status: EvalRunStatus;
  started_at: string | null;
  finished_at: string | null;
  hit_rate: number | null;
  mrr: number | null;
  answer_pass_rate: number | null;
  files_per_minute: number | null;
  error: string | null;
}

export function listEvalDatasets() {
  return request<EvalDataset[]>("/evaluation/datasets");
}

export function createEvalDataset(name: string, testset: File, corpus: File[]) {
  const body = new FormData();
  body.append("name", name);
  body.append("testset", testset);
  // FastAPI reads repeated parts as `list[UploadFile]`.
  corpus.forEach((file) => body.append("corpus", file));
  return request<EvalDataset>("/evaluation/datasets", { method: "POST", body });
}

export function deleteEvalDataset(id: string) {
  return request<void>(`/evaluation/datasets/${encodeURIComponent(id)}`, { method: "DELETE" });
}

export function listEvalRuns(limit = 50) {
  return request<EvalRunSummary[]>(`/evaluation/runs?limit=${limit}`);
}

export function startEvalRun(datasetId: string) {
  return request<EvalRun>("/evaluation/runs", {
    method: "POST",
    body: JSON.stringify({ dataset_id: datasetId }),
  });
}

export function getEvalRun(id: string) {
  return request<EvalRun>(`/evaluation/runs/${encodeURIComponent(id)}`);
}

export function cancelEvalRun(id: string) {
  return request<EvalRun>(`/evaluation/runs/${encodeURIComponent(id)}/cancel`, { method: "POST" });
}
